### DISCUSSION TITLE
error: 'projector_oct' undefined

### DISCUSSION THREAD

**Discussion by tomek18** [villekf/OMEGA#31] [Q&A]
Hi,
The OMEGA framework is an impressive tool for PET reconstruction and a lot can be learned while going through individual m files. 
Upon OMEGA installation (using Octave) and running install_mex I got messages ...."Implementations 1 & 4 built" and "Implementation 3 built". However, when testing the "lor_pixel_count_prepass.m" code an  error: "'projector_oct' undefined" appears although the _source_ folder contains .oct files. Any ideas why this happens?
Best regards,
Tom

**Comment by villekf** [comment]
Hi,

Are you on Windows or Linux? Does `projector_oct.oct` exist in the source folder?

**Comment by tomek18** [comment]
Hi,
I am on Windows and projector_oct.oct exists in the source folder!
Best regards,
Tom

**Reply by villekf** [reply]
As a clarification, are you running the `lor_pixel_count_prepass.m` on its own or part of e.g. `reconstructions_main`? Have you added the source-folder to the Octave path?

**Comment by tomek18** [comment]
Hi,
The path was set and I am running it on its own but taking care that all
"option" parameters are initalised/defined.
Best regards,
Tom

**Reply by villekf** [reply]
Does it work if you run it while inside the source-folder?

**Comment by tomek18** [comment]
What exactly do you mean by "inside the folder"? How would you suggest
trying to run the " [ lor, lor_orth, lor_vol] = projector_oct( Ny, Nx,
Nz,......"

**Reply by villekf** [reply]
What I mean is that the source-folder of OMEGA would be the current directory in Octave when you run the function.

**Comment by tomek18** [comment]
Hi,
Yes, it is, but unfortunately, the error is still the same. I will test the
code on another Windows system with the new Matlab.
Being able to run the code under Octave would be perfect, but it might be
that I did something wrong also during the installation process.
Thank you very much for your prompt reply and efforts. I will try to
discover what might be causing the problem and let you know if I find out
the reason, hopefully after the weekend when I will have some more time to
devote to the problem.

Best regards,
Tom

**Reply by villekf** [reply]
Just tested this myself and after some debugging, I believe I've found the reason. The previous update that introduced CT support, also broke implementation 1 for PET in Octave. This is because in `projector_oct.cpp` the function name is `projector_octCT`. If you don't need CT support at all, simply modify `DEFUN_DLD(projector_octCT, prhs, nargout, "projector_octCT") {` to `DEFUN_DLD(projector_oct, prhs, nargout, "projector_oct") {` and it should work after compiling the code again (I haven't managed to test this yet though). I'll fix this in master tomorrow. Testing after the update was, unfortunately, lacking on Octave with regards to PET.

**Comment by tomek18** [comment]
Hi,
Thank you for your efforts and explanation. I will try the suggested change
during the day.
Best regards,
Tom

**Comment by villekf** [comment]
This has now been fixed on master. I was able to successfully compute the reconstructions for implementation 1, though this should have affected implementation 4 and the `lor_pixel_count_prepass.m` you've used as well. You can also just download the fixed file [projector_oct.cpp](https://raw.githubusercontent.com/villekf/OMEGA/master/source/projector_oct.cpp) and run `install_mex` again. That should keep the CT part working as well.

Thank you for reporting this issue. I'll have to try to test the Octave execution better in the future.

**Comment by SomaKohsei** [comment]
HELLO.
I have a question here that may be relevant.
I am using octave5.2 in virtualBox (vGate v8.2). I used the sample ROOT file (https://zenodo.org/record/3526859#.ZC1P9HbP1nQ) and ran it with only the following 3 changes in gate_main.m

ASCII -> ROOT
implementation 2 -> 4
options.fpath = /path/to/GATE/output

Then I get the following error.
Is there any way to fix this?

Beginning LOR precomputation phase. This may take several minutes.
error: 'x_center' undefined near line 509 column 139
error: called from
    lor_pixel_count_prepass at line 506 column 39
    precompute_data at line 46 column 5
    gate_main at line 1760 column 4

thank you 
Kohsei Soma

**Comment by villekf** [comment]
`gate_main.m` should already use implementation 4 by default. Are you sure you didn't modify `options.projector_type`? I actually don't have a check for incorrect projector type at the beginning, only just before the reconstruction itself so you would not get any errors on that. If `options.projector_type` is set at 1, 2 or 3, then a quick fix should be to set `options.precompute = false`.

**Reply by SomaKohsei** [reply]
Dear villekf,

I am so sorry.
I mistakenly put 4 in projector_type. (Not implementation.)

I was able to get a reconstructed image!!

I have tried and failed with many software.
Your kindness will help me to keep trying!
Thanks for your wonderful software !!

Please let me ask you another question sometime.

thank you

**Reply by villekf** [reply]
Next release will print out an error in such cases.
